subprojects/execution/src/main/java/org/gradle/internal/execution/impl/OutputsCleaner.java

```diff
@@ -0,0 +1,98 @@
+/*
+ * Copyright 2019 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.gradle.internal.execution.impl;
+
+import com.google.common.collect.Ordering;
+import org.gradle.internal.file.FileType;
+import org.gradle.internal.fingerprint.FileCollectionFingerprint;
+import org.gradle.internal.fingerprint.FileSystemLocationFingerprint;
+
+import javax.annotation.Nullable;
+import java.io.File;
+import java.util.Map;
+import java.util.PriorityQueue;
+import java.util.function.Predicate;
+
+public class OutputsCleaner {
```

Comment:
Why is this so complicated? Why don't we use `GFileUtils.cleanDirectory()` and friends here?

Reply:
Since we need to be able to cleanup the output directory when there are overlapping outputs. Most of this code is from `SimpleStaleClassCleaner`.

Reply:
I wonder if we need to put in some extra measures so we don't clean up outputs outside of the currently configured outputs. If the outputs are configured to be in another location, then the task history will still point to the old outputs somewhere else. This cleanup step would clean up the old outputs, even though they are not anymore within the currently configured output directories.

Do you think we should prevent this or clean up the old outputs in the old place?

Reply:
If the old location is owned by the build, then I'd argue we should clean it up. If I change `JavaCompile.destinationDir` it would be nice to see the `.class` files from the old location go. We'd even remove the old `destinationDir` itself, right? Because it's not among the protected directories. That's nice.